File: programmers/kakao/캐시.java
package A1;

import java.io.*;
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
    	if(cacheSize == 0) {
    		return cities.length * 5;
    	}
        int answer = 0;
        ArrayList<String> cache = new ArrayList<>();
        int idx;
        for(int i = 0; i < cities.length; ++i) {
        	String city = cities[i].toLowerCase();
        	if((idx = find(cache, city)) != -1){
        		cache.remove(idx);
        		answer++;
        	}else {
        		if(cache.size() >= cacheSize) {
        			cache.remove(0);
        		}
        		answer += 5;
        	}
        	cache.add(city);
        }
        return answer;
    }
    int find(ArrayList<String> cache, String city) {
    	for(int i = 0; i < cache.size(); ++i) {
    		if(cache.get(i).equals(city)) {
    			return i;
    		}
    	}
    	return -1;
    }
}

public class 캐시 {
	public static void main(String[] args) {
		Solution sol = new Solution();
		int cacheSize = 3;
		String[] cities = {"Jeju", "Pangyo", "Seoul", "NewYork", "LA", "Jeju", "Pangyo", "Seoul", "NewYork", "LA"};
		System.out.println(sol.solution(cacheSize, cities));
	}
}

File: programmers/js/new/level2/캐시.js
function solution(cacheSize, cities) {
    var answer = 0;
    const cache = [];
    for (let i = 0; i < cities.length; ++i) {
        const city = cities[i].toLowerCase();
        if (findInCache(cache, cacheSize, city)) {
            answer++;
        } else {
            answer += 5;
            insertInCache(cache, cacheSize, city);
        }
    }
    return answer;
}

function findInCache(cache, cacheSize, city) {
    let findIdx = -1;
    for (let i = 0; i < cacheSize; ++i) {
        if (cache[i] === city) {
            findIdx = i;
            break;
        }
    }
    if (findIdx >= 0) {
        cache.splice(findIdx, 1);
        cache.unshift(city);
        return true;
    }
    return false;
}

function insertInCache(cache, cacheSize, city) {
    if (cache.length < cacheSize) {
        cache.unshift(city);
    } else {
        cache.pop();
        cache.unshift(city);
    }
}

console.log(solution(2, ["Jeju", "Pangyo", "NewYork", "newyork"]));
